### PR TITLE
Convert compatible pypi package names to the ones used in pypi

### PR DESCRIPTION
**Pull request by qtomlinson**
Pypi package management system treats ".", "-", and "_" as the same. For example, the following three urls yield the same component, backports.ssl_match_hostname.
https://pypi.org/project/backports.ssl.match.hostname/
https://pypi.org/project/backports_ssl_match_hostname/
https://pypi.org/project/backports-ssl-match-hostname/

During crawling, package coordinates internally stored in the harvest store are consistent with the package management repository.  Harvested information can only be retrieved based on the same standardized coordinates.

To allow retrieval of pypi package information via compatible names, added converting input package names to the ones used in the python package index. This conversion is done at all retrieval api entry points, and by querying pypi index (same as in crawler).  The query result is cached for later reuse and set to expire after configured life span.

Tasks: https://github.com/clearlydefined/service/issues/860

### CODE REVIEW

**Comment by qtomlinson** [comment]
@nellshamrell This pull request is dependent on https://github.com/clearlydefined/service/pull/891

**Comment by nellshamrell** [comment]
hmm...just checked out this branch locally and the tests passed fine. Seeing if there is something up in the build system.

**Comment by nellshamrell** [comment]
Actually, I had the wrong branch checked out. Trying again locally.

**Comment by nellshamrell** [comment]
yep, looks like it's failing locally as well

**Comment by qtomlinson** [comment]
Thanks for the feedback.  I will try to set up another clean branch and have a look.

**Comment by nellshamrell** [comment]
Looks like the failure started in [this commit](https://github.com/clearlydefined/service/pull/893/commits/3dfbcfa192180e791cb43f687f89e516a484f985)

**Comment by nellshamrell** [comment]
I think I know what it is :) 

"test/lib/entityCoordintates.js" should be "test/lib/entityCoordinates.js"

(No worries! I do stuff like that all the time!)

**Comment by qtomlinson** [comment]
Thanks for catching the naming mistake.  I will fix that.

**Comment by nellshamrell** [comment]
Still have a failure - but I think I know what's up. Confirming.

File: lib/coordinatesMapper.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+class CoordinatesMapper {
+
+  constructor(mappers = defaultMappers, cache = defaultcache) {
+    this.mappers = mappers
+    this.cache = cache
+  }
+
+  async map(coordinates) {
+    const mapper = this.mappers[coordinates?.provider]
```
Comment (nellshamrell) [inline lib/coordinatesMapper.js:19]: There's something going on with this line that's causing the `SyntaxError: Unexpected token '.'` (which I only found by running the mocha tests in debug mode in VS Code)

**Comment by nellshamrell** [comment]
You can isolate the error by running just the coordinatesMapper test with `npx nyc mocha test/lib/coordinatesMapper.js`

**Comment by nellshamrell** [comment]
I think it's still not able to process the optional chaining operator when running tests, even though eslint is updated

**Comment by nellshamrell** [comment]
Ah, that operator is a feature of Node 14. Checking what version of node we are using...

**Comment by nellshamrell** [comment]
Yep, we're using Node 12 in our Docker builds. We can look into updating to Node 14 (might be risky, but likely worth doing sometime soon) or rewrite the code to be compatible with Node 12.

File: lib/coordinatesMapper.js (hunk 1 above)
Comment (qtomlinson) [reply lib/coordinatesMapper.js:19]: I only see this error when parserOptions.ecmaVersion is changed to 9 in .eslintrc.json.  This should be addressed by my eslint upgrade commit.  Does eslint in build pipeline use config in formats other than json?

**Comment by qtomlinson** [comment]
I will change to avoid using optional chaining in the mean time.

**Comment by qtomlinson** [comment]
Just noticed Node.js installed in our build pipeline is 12.x as well.  See azure-pipelines.yml

**Comment by nellshamrell** [comment]
Update to Node 14 is complete! Reviewing this again.

**Comment by nellshamrell** [comment]
This overall looks great! I'm going to do some final testing tomorrow then should be able to deploy it :) Thank you so much for this outstanding work @qtomlinson!

**Comment by qtomlinson** [comment]
@nellshamrell Thanks for the feedback! A couple of points to be discussed:
1.	Naming of Class: CoordinatesMapper, it maps/converts the user specified coordinates to the standard one used in the package repository.  Do you have suggestion for naming?
2.	Package location for the added classes: CoordinatesMapper and PypiCoordinatesMapper.  CoordinatesMapper is responsible for caching and delegate to the provider specific classes for coordinates retrieval.  The classes are currently in /lib.  Is there a better code location for them?
3.	The default for life span of mapped coordinates in cache is currently set to be 1 day.   Should this default changed to a longer period?

**Comment by nellshamrell** [comment]
Hi @qtomlinson!

1. I think "CoordinatesMapper" is fine as a name
2. Being in /lib is fine
3. The cache being one day is also fine

Great work!

**Review by nellshamrell — APPROVED** [review]
This is great! I will deploy this tomorrow :)

**Pull request merged**